walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -23,41 +23,74 @@
-      Collection<String> meetingAttendees = request.getAttendees();
-      if (Collections.disjoint(eventAttendees, meetingAttendees)) {
+      if (Collections.disjoint(eventAttendees, mandatoryMeetingAttendees)) {
+        if (!Collections.disjoint(eventAttendees, optionalMeetingAttendees)) {
+          optionalEvents.add(event);
+        }
         continue;
       }
-      TimeRange eventTime = event.getWhen();
-      for (TimeRange meetingTime : meetingTimes) {
-        if (eventTime.overlaps(meetingTime)) {
-          meetingTimes.remove(meetingTime);
-          if (meetingTime.start() < eventTime.start()) {
-            TimeRange before = TimeRange.fromStartEnd(meetingTime.start(), eventTime.start(), false);
-            if (before.duration() >= request.getDuration()) {
-              meetingTimes.add(before);
-            }
-          }
-          if (eventTime.end() < meetingTime.end()) {
-            TimeRange after = TimeRange.fromStartEnd(eventTime.end(), meetingTime.end(), false);
-            if (after.duration() >= request.getDuration()) {
-              meetingTimes.add(after);
-            }
-          }
-        }
+      meetingTimes = splitMeetingTimes(event, meetingTimes, meetingDuration);
+      if (meetingTimes.isEmpty()) {
+        break;
       }
     }
+
+    Collection<TimeRange> optionalMeetingTimes = new TreeSet<>(TimeRange.ORDER_BY_START);
+    optionalMeetingTimes.addAll(meetingTimes);
+    for (Event optionalEvent : optionalEvents) {
+      optionalMeetingTimes = splitMeetingTimes(optionalEvent, optionalMeetingTimes, meetingDuration);
+      if (optionalMeetingTimes.isEmpty()) {
+        break;
+      }
+    }
+
+    if (mandatoryMeetingAttendees.isEmpty()) {
+      meetingTimes = optionalMeetingTimes;
+    } else {
+      meetingTimes = optionalMeetingTimes.isEmpty() ? meetingTimes : optionalMeetingTimes;
+    }
+
     Collection<TimeRange> finalMeetingTimes = new ArrayList<>();
     finalMeetingTimes.addAll(meetingTimes);
     return finalMeetingTimes;
   }
+
+  public Collection<TimeRange> splitMeetingTimes(Event event, Collection<TimeRange> meetingTimes, long meetingDuration) {
+    Collection<TimeRange> newMeetingTimes = new TreeSet<>(TimeRange.ORDER_BY_START);
+    TimeRange eventTime = event.getWhen();
+    for (TimeRange meetingTime : meetingTimes) {
+      // TODO: Reduce if nesting complexity.
+      if (eventTime.overlaps(meetingTime)) {
+        if (meetingTime.start() < eventTime.start()) {
+          TimeRange before = TimeRange.fromStartEnd(meetingTime.start(), eventTime.start(), false);
+          if (before.duration() >= meetingDuration) {
+            newMeetingTimes.add(before);
+          }
+        }
+        if (eventTime.end() < meetingTime.end()) {
+          TimeRange after = TimeRange.fromStartEnd(eventTime.end(), meetingTime.end(), false);
+          if (after.duration() >= meetingDuration) {
+            newMeetingTimes.add(after);
+          }
+        }
```

Comment:
You have 2 piece of code here (lines 79-83 and lines 84-88) that are almost identical.  This seems like another good case to factor out a function.